### PR TITLE
[WIP] util: add OOM hook

### PR DESCRIPTION
**Pull request by Hijiao**
Signed-off-by: Max <xiahaijiao@hotmail.com>

## What have you changed? (mandatory)
Add an OOM hook which collects backtrace info when OOM like `panic_hook`

## What are the type of the changes? (mandatory)

The currently defined types are listed below, please pick one of the types for this PR by removing the others:
- Improvement (non-breaking change which is an improvement to an existing feature)

## How has this PR been tested? (mandatory)
No, it waits for CI test.

## Does this PR affect documentation (docs) update? (mandatory)

No

## Does this PR affect tidb-ansible update? (mandatory)
No

### CODE REVIEW

**Comment by brson** [comment]
It would be nice to come up with a way to test this automatically to know that it will work when needed. Likewise the panic hook if it's not already tested.

**Comment by brson** [comment]
I [filed a bug against std](https://github.com/rust-lang/rust/issues/56965) asking for documentation clarifications for these oom hooks.

File: src/util/mod.rs [hunk 1]
```diff
@@ -521,6 +529,34 @@ pub fn set_exit_hook(
...
+fn set_oom_hook() {
+    use std::alloc::{set_alloc_error_hook, take_alloc_error_hook, Layout};
+    use std::process;
+
+    fn hook(layout: Layout) {
+        let origin_alloc_error_hook = take_alloc_error_hook();
+        if log_enabled!(::log::LogLevel::Error) {
+            let thread = thread::current();
+            let name = thread.name().unwrap_or("<unnamed>");
+            let bt = ::backtrace::Backtrace::new();
```
Comment (overvenus) [inline src/util/mod.rs]: Does it work if we alloc it ahead (in other stack, the `set_oom_hook`) ?

**Comment by brson** [comment]
@overvenus Storing some kind of "emergency" allocation is a good idea, though in this where one wants to create a `Backtace` value, it would need to be integrated into the backtrace library.

The code for creating `Backtrace` is pretty simple: https://github.com/alexcrichton/backtrace-rs/blob/master/src/capture.rs#L20

I see three sources of allocations: the vec of `BacktraceFrames`, which is optional, and the vecs of `BacktraceSymbol` and `PathBuf`, which are optional, but which one almost certainly wants.

To use 'emergency' vecs in this situation you'd have a few considerations:

- They need to be initialized at some point
- They need to be overallocated
- Backtrace needs to be modified to know when and when not to use the emergency allocation
- It needs to know not to let the emergency allocations grow
- It needs to consider concurrency - e.g. you might want the oom to take a global lock, allocate the backtrace, print it, drop the backtrace, and release the emergency allocations back to their global locations.
- It might want to live upstream in the backtrace crate, since others are going to want to take backtraces during oom.

I suspect it's not worth the effort to try to make the convenient `Backtrace` type compatible with OOM, but rather to learn from what it's doing to traverse the stack frame, while logging the frame information without allocation.

Looking at the `Backtrace` implementation there are basically two steps:

- calling [`trace`](https://github.com/alexcrichton/backtrace-rs/blob/master/src/capture.rs#L99) to get callbacks for every frame
- calling  [`resolve`](https://github.com/alexcrichton/backtrace-rs/blob/master/src/capture.rs#L143) on every frame to get the symbol names

The backtracing and symbol resolution code is really complicated, but it looks to me that it's designed to work with out allocation (e.g. in no_std).

A non-allocating, symbol-resolving, demangling, stack frame iterator _looks_ like something relatively easy to write by following the code for `Backtrace`, but it would probably be clever to ask @alexcrichton if that makes sense. There may already be such code somewhere that we could just take, or if you write one maybe it would be a useful addition to the backtrace crate.

**Comment by brson** [comment]
On the other hand, after writing that, just giving `Backtrace` a new constructor containing a giant handful of cached, pre-allocated vecs is kind of attractive, since that would be an easy hack.

**Comment by brson** [comment]
I'm curious what you think the next steps here are @Hijiao.

I still don't have a firm idea of what to do, but offhand I think this would be a good starting point:

- Create a test case that checks that the oom handler doesn't oom. It might look like: link to a global test allocator that in turn calls jemalloc. Have the test case 'turn off' the allocator per thread, try to allocate something, and confirm that it doesn't keep ooming. To avoid the runtime-imposed abort you could catch a panic with a particular value to indicate success. Maybe this test has the rotating file logger turned off, and a second one has it turned on. The global test allocator would _prabably_ want to override the malloc symbols like jemalloc does, to catch any allocations in C code.

  This test will presumably fail.

- Get both loggers tikv uses to definitely work without allocating.

- Audit libbacktrace (or whatever symbolification library we link to on linux) to see whether it looks like it calls malloc on our code path.

- Probably write a backtrace iterator or implementation of `Backtrace` that doesn't allocate.

An alternative might be to just _hope_ the allocations work - e.g. even if one allocation fails, the next might not - and abort on recursive oom if it doesn't work. This would only be useful if the oom handler often does work To give it a better chance of working, maybe there are caches in the system that can be freed right before aborting on oom.

**Comment by Hijiao** [comment]
@alexcrichton  Thanks for your help, the underlying libbacktrace does allocate. I add a `dumps_emergency` method for `Backtrace` that directly writes `backtrace::{trace, resolve}` results into a passed pre-allocated buf, you can see the code [here](https://github.com/Hijiao/backtrace-rs/commit/833551ee5262dc711a685ebf50aa5e2242d49420). After carefully test, `jemalloc stats` shows that, the malloc request increases about 30.

**Comment by Hijiao** [comment]
@brson Sorry for later replay, those days I was catching up on knowledge about the allocation which is unfamiliar for me. I used `strace` and `jeprof` to try to find out the detail of allocation in underlying `libbacktrace` with system allocator or jemalloc allocator. As a result, it does allocate, but until now I don't know clearly how it allocates the new memory.

According to my test [code](https://github.com/Hijiao/tikv/commit/8ee36be82e99396a99d94aabb5cf3ed189172245) , `jemalloc stats` tells that, when use jemalloc on Linux platform, the `nrequests`([link](http://jemalloc.net/jemalloc.3.html#mallctl_namespace))  of `backtrace_dumps_test` increases about 30 than `backtrace_no_dumps_test`, but while on MacOS, the `nrequests` are same. My MacBook fools me at the beginning. 😂

In my original plan:
1. dump backtrace info into a pre-allocated buf.
2. flush the buf to stderr instead of tikv log file. 

but now I am stuck in the step 1 because of the allocation of underlying `libbacktrace`. Thanks to your kind advice, in the next, I will try the global allocator. 

Thank you, Brian!

**Comment by overvenus** [comment]
> Another way to do the allocation ahead of time would be to just copy-paste the Backtrace type from the backtrace crate and modify it to use a fixed number and capacity of vectors. Call it OomBacktrace. That would be pre-instantiated as you suggest when setting the hook. This wouldn't work if symbol resolution allocates.

Agree, it does not work because libbacktrace itself allocates.

> Another way to do the allocation ahead of time would be to override the global allocator as a wrapper around jemalloc plus an oom cache, maybe a simpler version of malloc backed by a pre-allocated vec. Then when it first ooms it can swap jemalloc for the pre-allocated malloc, generate whatever allocations the oom handler needs, then abort. If the native symbol resolution allocates this may not work or may only work on platforms where malloc can be overridden reliably.

It's a good idea, though writing a custom alloctor is a tricky task. I'm not sure if we can use rust's _global_allocator_ directly(does it handle `malloc` from C libraries?).

**Comment by alexcrichton** [comment]
@brson 

> Do you know of any other projects that are doing backtraces without allocation, especially with symbol resolution?

AFAIK no. For `backtrace-sys` we also explicitly disabled the OOM-safe allocator which uses mmap instead of malloc/free because it was so slow. I don't think that `backtrace-sys`/`backtrace` are in a suitable position currently to perform backtraces and resolve symbols with zero allocations, safe for an OOM event. 

The best way forward may be to work with `gimli` explicitly which I believe also has `no_std` modes, although that would likely require significant investment as the infrastructure isn't really there yet to work correctly. In theory though process startup could execute the one-time "parse everything and make internal data structures" and then stack unwinding would be an allocation-free operation afterwards.

**Comment by brson** [comment]
> For `backtrace-sys` we also explicitly disabled the OOM-safe allocator which uses mmap instead of malloc/free because it was so slow. I don't think that `backtrace-sys`/`backtrace` are in a suitable position currently to perform backtraces and resolve symbols with zero allocations, safe for an OOM event.

Oh that's disappointing. I didn't realize that just the basic backtrace allocates.

> The best way forward may be to work with gimli explicitly which I believe also has no_std modes, although that would likely require significant investment as the infrastructure isn't really there yet to work correctly. In theory though process startup could execute the one-time "parse everything and make internal data structures" and then stack unwinding would be an allocation-free operation afterwards.

But that sounds fun @Hijiao. The world needs a zero-allocation backtrace parser.

Good luck in your investigations. I'm happy to (try to) help if you get stuck.

**Comment by Hijiao** [comment]
@brson  Thanks for your kindness, but this task is not worth the much effect, I decide to remain it on limbo.😅

**Pull request closed without merging**